ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueCommitting.scala

```diff
@@ -146,6 +142,157 @@ object KeyValueCommitting {
     }
   }
 
+  private def processPackageUpload(
+      entryId: DamlLogEntryId,
+      recordTime: Timestamp,
+      packageUploadEntry: DamlPackageUploadEntry,
+      inputState: Map[DamlStateKey, Option[DamlStateValue]]
+  ): (DamlLogEntry, Map[DamlStateKey, DamlStateValue]) = {
+    val submissionId = packageUploadEntry.getSubmissionId
+    def tracelog(msg: String): Unit =
+      logger.trace(
+        s"processPackageUpload[entryId=${prettyEntryId(entryId)}, submId=$submissionId]: $msg")
+
+    val archives = packageUploadEntry.getArchivesList.asScala
+    val packagesValidityResult =
+      archives.foldLeft[(Boolean, String)]((true, ""))(
+        (acc, archive) =>
+          if (archive.getPayload.isEmpty) (false, acc._2 ++ s"empty package '${archive.getHash}';")
+          else acc)
+
+    if (packagesValidityResult._1) {
+      val filteredArchives = archives
+        .filter(
+          archive =>
+            inputState(
+              DamlStateKey.newBuilder
+                .setPackageId(archive.getHash)
+                .build).isEmpty
+        )
+      tracelog(s"""Packages: ${archives
+        .map(_.getHash)
+        .mkString(",")} committed.""")
+
+      (
+        DamlLogEntry.newBuilder
+          .setRecordTime(buildTimestamp(recordTime))
+          .setPackageUploadEntry(
+            DamlPackageUploadEntry.newBuilder
+              .setSubmissionId(submissionId)
+              .addAllArchives(filteredArchives.asJava)
+              .setSourceDescription(packageUploadEntry.getSourceDescription)
+              .setParticipantId(packageUploadEntry.getParticipantId)
+              .build
+          )
+          .build,
+        filteredArchives
+          .map(
+            archive =>
+              (
+                DamlStateKey.newBuilder.setPackageId(archive.getHash).build,
+                DamlStateValue.newBuilder.setArchive(archive).build
+            )
+          )(breakOut)
+      )
+    } else {
+      tracelog(s"Packages upload failed, invalid package submitted")
```

Comment:
log what invalid package it was and why it was invalid.

Reply:
Done